src/common/function/FunctionManager.cpp

```diff
@@ -577,10 +582,78 @@ FunctionManager::FunctionManager() {
         }
         case Value::Type::INT:
         case Value::Type::FLOAT: {
-          if (args.size() == 2) {
+          if (args.size() >= 2) {
             if (args[1].get().type() == Value::Type::INT) {
+              auto val = args[0].get().getFloat();
               auto decimal = args[1].get().getInt();
-              return std::round(args[0].get().getFloat() * pow(10, decimal)) / pow(10, decimal);
+              auto factor = pow(10.0, decimal);
+
+              string mode = "half_up";
+              if (args.size() == 3 && args[2].get().type() == Value::Type::STRING) {
+                mode = args[2].get().getStr();
+              }
+              if (mode == "up") {
+                auto roundedVal = std::round(val * factor) / factor;
+
+                if (val < 0) {
+                  if (roundedVal > val) {
+                    roundedVal -= 1.0 / factor;
+                  }
+                } else if (val > 0) {
+                  if (roundedVal < val) {
+                    roundedVal += 1.0 / factor;
+                  }
+                }
+                return roundedVal;
+              } else if (mode == "down") {
+                auto roundedVal = std::round(val * factor) / factor;
+                if (val < 0) {
+                  if (roundedVal < val) {
+                    roundedVal += 1.0 / factor;
+                  }
+                } else if (val > 0) {
+                  if (roundedVal > val) {
+                    roundedVal -= 1.0 / factor;
+                  }
+                }
+                return roundedVal;
+              } else if (mode == "ceiling") {
+                return std::ceil(val * factor) / factor;
+              } else if (mode == "floor") {
+                return std::floor(val * factor) / factor;
+              } else if (mode == "half_up") {
+                return std::round(val * factor) / factor;
+              } else if (mode == "half_down") {
+                auto next_largest = std::ceil(val * factor) / factor;
+                auto next_smallest = std::floor(val * factor) / factor;
+                if (std::fabs((val - next_smallest) - (next_largest - val)) < FLT_EPSILON) {
```

Comment:
There may be a more efficient way to determine tie.
Fwiw, static_cast<int>(val * factor)&0x0111.

Reply:
Thank you for your excellent suggestion! I have updated the logic not to calculate both of the largest and smallest.